msprime/intervals.py

```diff
@@ -229,26 +341,111 @@ def uniform_map(cls, length, rate, num_loci=None):
+        only the second (Position) and third (Rate) columns are read: any
+        further columns are ignored. If the first starting position is not equal
+        to zero, then a zero-recombination region is inserted at the start of
+        the chromosome.
 
         .. warning::
             This method is deprecated, use the module-level
-            :func:`read_hapmap` function instead.
+            :func:`read_hapmap` function instead, which reads the genetic position
+            in centiMorgans from the file, rather than using the rate column.
 
         :param str filename: The name of the file to be parsed. This may be
-            in plain text or gzipped plain text.
+            in plain text or gzipped plain text. It may also be a file object.
+        :param str sequence_length: The total length of the map. If ``None``, then
+            assume it is the final physical position in the file. Otherwise pad the
+            map with a zero rate region between the final position in the file and
+            the value specified here.
         :return: A RecombinationMap object.
         """
         warnings.warn(
             "RecombinationMap.read_hapmap() is deprecated. "
             "Use msprime.read_hapmap() instead.",
             FutureWarning,
         )
-        rate_map = read_hapmap(filename)
-        rate = np.append(rate_map.rate, 0)
-        return cls(rate_map.position, rate, map_start=rate_map.map_start)
+        rownames, position, rate = np.loadtxt(
+            filename,
+            skiprows=1,
+            dtype=[("chrom", object), ("position", int), ("rate", float)],
+            usecols=(0, 1, 2),
+            unpack=True,
+        )
+        if len(rownames) == 0:
+            raise ValueError("Empty hapmap file")
+
+        if len(np.unique(rownames)) > 1:
+            warnings.warn(
+                f"Different chromosome names found in the first column of {filename}. "
+                "Check that the file contains data for a single chromosome."
+            )
+        name = rownames[0]
+        # Rate is expressed in centimorgans per megabase: convert to per-base rates
+        rate *= 1e-8
+
+        start = position[0]
+        end = position[-1]
+        if start != 0:
+            position = np.insert(position, 0, 0)
+            rate = np.insert(rate, 0, 0)
+        if rate[-1] != 0:
+            raise ValueError(
+                "The last rate provided in the recombination map must be zero"
+            )
+        if sequence_length is not None:
+            if sequence_length < end:
+                raise ValueError(
+                    f"Cannot specify a sequence_length < {end} (the last position)"
+                )
+            elif sequence_length > end:
+                position = np.append(position, sequence_length)
+                rate = np.append(rate, 0)
+        return cls(position, rate, name=name, start_position=start, end_position=end)
+
+    def write_hapmap(self, filename):
```

Comment:
This is handy, but lets put it into a module function, like ``read_hapmap``. The file argument should be a fileobj rather than a filename too.

Reply:
This is only for "old" maps, to convert them to the new style, so I didn't think it was worth implementing functionality to write out the RateMaps. I guess I could generalise it though.